code/__HELPERS/datastructs/armor.dm

```diff
@@ -1,3 +1,234 @@
-// tbd
+/**
+ * armor datums for holding values
+ * globally cached.
+ */
+/datum/armor
+	/// just for vv
+	var/name
+	
+	var/melee = 0
+	var/melee_tier = ARMOR_TIER_DEFAULT
+	var/melee_soak = 0
+	var/bullet = 0
+	var/bullet_tier = ARMOR_TIER_DEFAULT
+	var/bullet_soak = 0
+	var/laser = 0
+	var/laser_tier = ARMOR_TIER_DEFAULT
+	var/laser_soak = 0
+	var/energy = 0
+	var/bomb = 0
+	var/bio = 0
+	var/rad = 0
 
-// /proc/fetch_armor_struct
+/datum/armor/New(list/from_values)
+	if(from_values)
+		from_list(from_values)
+	name = to_name()
+
+/datum/armor/vv_edit_var(var_name, var_value, mass_edit, raw_edit)
+	if(var_name in global.armor_enums)
+		return FALSE // no.
+	return ..()
+
+/datum/armor/proc/from_list(list/values)
+	#define UNPACK_OR(var, key, default) ##var = isnull(values[key])? default : values[key]
+	UNPACK_OR(melee, ARMOR_MELEE, 0)
+	UNPACK_OR(melee_tier, ARMOR_MELEE_TIER, ARMOR_TIER_DEFAULT)
+	UNPACK_OR(melee_soak, ARMOR_MELEE_SOAK, 0)
+	UNPACK_OR(bullet, ARMOR_BULLET, 0)
+	UNPACK_OR(bullet_tier, ARMOR_BULLET_TIER, ARMOR_TIER_DEFAULT)
+	UNPACK_OR(bullet_soak, ARMOR_BULLET_SOAK, 0)
+	UNPACK_OR(laser, ARMOR_LASER, 0)
+	UNPACK_OR(laser_tier, ARMOR_LASER_TIER, ARMOR_TIER_DEFAULT)
+	UNPACK_OR(laser_soak, ARMOR_LASER_SOAK, 0)
+	UNPACK_OR(energy, ARMOR_ENERGY, 0)
+	UNPACK_OR(bomb, ARMOR_BOMB, 0)
+	UNPACK_OR(bio, ARMOR_BIO, 0)
+	UNPACK_OR(rad, ARMOR_RAD, 0)
+	#undef UNPACK_OR
+
+/datum/armor/proc/to_list()
+	return list(
+		ARMOR_MELEE = melee,
+		ARMOR_MELEE_TIER = melee_tier,
+		ARMOR_MELEE_SOAK = melee_soak,
+		ARMOR_BULLET = bullet,
+		ARMOR_BULLET_TIER = bullet_tier,
+		ARMOR_BULLET_SOAK = bullet_soak,
+		ARMOR_LASER = laser,
+		ARMOR_LASER_TIER = laser_tier,
+		ARMOR_LASER_SOAK = laser_soak,
+		ARMOR_ENERGY = energy,
+		ARMOR_BOMB = bomb,
+		ARMOR_BIO = bio,
+		ARMOR_RAD = rad,
+	)
+
+/datum/armor/proc/to_name()
+	return jointext(list(
+		"[melee]@[melee_tier]-[melee_soak]",
+		"[bullet]@[bullet_tier]-[bullet_soak]",
+		"[laser]@[laser_tier]-[laser_soak]",
+		"[energy]",
+		"[bomb]",
+		"[bio]",
+		"[rad]",
+	), "-")
+
+/datum/armor/proc/raw(flag)
+	switch(flag)
+		if(ARMOR_MELEE)
+			return melee
+		if(ARMOR_MELEE_SOAK)
+			return melee_soak
+		if(ARMOR_MELEE_TIER)
+			return melee_tier
+		if(ARMOR_BULLET)
+			return bullet
+		if(ARMOR_BULLET_SOAK)
+			return bullet_soak
+		if(ARMOR_BULLET_TIER)
+			return bullet_tier
+		if(ARMOR_LASER)
+			return laser
+		if(ARMOR_LASER_SOAK)
+			return laser_soak
+		if(ARMOR_ENERGY)
+			return energy
+		if(ARMOR_BOMB)
+			return bomb
+		if(ARMOR_BIO)
+			return bio
+		if(ARMOR_RAD)
+			return rad
+
+/datum/armor/proc/mitigation(flag, tier = ARMOR_TIER_DEFAULT)
+	switch(flag)
```

Comment:
shouldnt there be an else here to cover all other cases?

Reply:
nope, if anything it should probably CRASH if someone is using the improper enums, as the only stuff that should go into the proc are the enums.